experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,198 @@
+---
+title: Building and Deploying a Rust library on Android using JNA
+layout: text
+---
+
+# Building and Deploying a Rust library on Android
+
+In our previous post we built a Rust library, and linked to it on [Android using JNI](https://mozilla.github.io/firefox-browser-architecture/experiments/2017-09-21-rust-on-android.html). This post builds on that by showing how to build the same app using JNA instead of JNI.
+
+# Why JNA?
+
+With JNI, for every `extern "C"` function that we created for our FFI, we had to create a mirror function in Rust for JNI to access. This involved some complicated function name construction and forced us to duplicate large chunks of code. JNA does away with all of that. You create a single interface that declares your FFI functions in Java and that's about it. It removes a large chunk of the headache and it is what we have moved to using here at Mozilla for our Android FFI.
```

Comment:
Might be worth mentioning that it does so at some perf cost - which, anecdotally, could be significant depending on the flow of stuff over this channel.

Reply:
Good idea. I'll mention a little bit here and link to https://github.com/java-native-access/jna/blob/master/www/FrequentlyAskedQuestions.md#how-does-jna-performance-compare-to-custom-jni which goes into more detail.